evm/src/generation/state.rs

```diff
@@ -93,6 +99,35 @@ impl<F: Field> GenerationState<F> {
         self.state_key_to_address.insert(state_key, address);
     }
 
+    /// Observe the given code hash and store the associated code.
+    /// When called, the code corresponding to `codehash` should be stored in the return data.
+    pub fn observe_contract(&mut self, codehash: H256) {
+        if self.inputs.contract_code.contains_key(&codehash) {
+            return; // Return early if the code hash has already been observed.
+        }
+
+        let ctx = self.registers.context;
+        let base_address = MemoryAddress::new(ctx, Segment::Returndata, 0);
+        let returndata_size_addr = MemoryAddress::new(
+            ctx,
+            Segment::ContextMetadata,
+            ContextMetadata::ReturndataSize as usize,
+        );
+        let returndata_size = self.memory.get(returndata_size_addr).as_usize();
+        let code = (0..returndata_size)
+            .map(|i| {
+                let address = MemoryAddress {
+                    virt: base_address.virt.saturating_add(i),
+                    ..base_address
+                };
+                let val = self.memory.get(address);
+                val.as_u32() as u8
+            })
+            .collect::<Vec<_>>();
+
```

Comment:
Maybe we should re-hash the code and check that it matches just to be safe? Could be debug-only